### PR TITLE
RoP: CODE of conduct Rule Integration

### PR DESCRIPTION
**Pull request by ajinkyapatil8190**
This pull request addresses The addition of the Code of conduct rule in the Fosstars check.

Checking if the code of conduct file is present in the repository or not and also checking if contributor covenant  text is present or not in the code of conduct file.

### CODE REVIEW

File: src/test/shell/tool/github/OssRulesOfPlayRatingMarkdownFormatter.config.yml [hunk 1]
```diff
@@ -17,4 +17,7 @@ ruleIds:
   rl-reuse_tool-2: If a project has a LICENSES folder with licenses
   rl-reuse_tool-3: If a project is registered in REUSE
   rl-reuse_tool-4: If a project is compliant with REUSE rules
-  rl-security_policy-1: If an open-source project has a security policy
\ No newline at end of file
+  rl-security_policy-1: If an open-source project has a security policy
+  rl-security_policy-1: If an open-source project has a security policy
```
Comment (sourabhsparkala) [inline src/test/shell/tool/github/OssRulesOfPlayRatingMarkdownFormatter.config.yml:21]: There are repeated lines

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfo.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * This data provider gathers info about project's code of conduct.
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfo.java:30]: Could you please elaborate, what Code of Conduct is? Maybe add a URL reference?

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/ContributingGuidelineInfo.java [hunk 3]
```diff
@@ -46,7 +46,7 @@
    * A list of paths to contributing guidelines.
    */
   private final List<String> knownContributingGuidelineFiles = new ArrayList<>();
-
+
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/data/github/ContributingGuidelineInfo.java:49]: Please remove this trailing space.

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/ContributingGuidelineInfo.java [hunk 4]
```diff
@@ -83,6 +83,7 @@ public ContributingGuidelineInfo knownContributingGuidelineFiles(String... files
     return this;
   }
 
+
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/data/github/ContributingGuidelineInfo.java:86]: Please remove this extra line.

**Review by sourabhsparkala — CHANGES_REQUESTED** [review]
There are test failures, could you have a look? https://github.com/SAP/fosstars-rating-core/pull/693/checks?check_run_id=3472217753#step:4:1958

Thanks
Sourabh

**Comment by ajinkyapatil8190** [comment]
> There are test failures, could you have a look? https://github.com/SAP/fosstars-rating-core/pull/693/checks?check_run_id=3472217753#step:4:1958
> 
> Thanks
> Sourabh

Hi @sourabhsparkala - I did the changes and now all the tests are passing, all the checkstyles are correct.

Please let me know for anymore changes from my end. 
Regards,
Ajinkya.

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/ContributingGuidelineInfo.java [hunk 5]
```diff
@@ -46,7 +46,6 @@
    * A list of paths to contributing guidelines.
    */
   private final List<String> knownContributingGuidelineFiles = new ArrayList<>();
-
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/data/github/ContributingGuidelineInfo.java:49]: Please do not remove this line

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfo.java [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+  /**
+   * Extract info about a code of conduct guideline.
+   *
+   * @param content Content of the code of conduct guideline.
+   * @return A value of {@link OssFeatures#HAS_REQUIRED_TEXT_IN_CODE_OF_CONDUCT_GUIDELINE}.
+   */
+  private Value<Boolean> infoAboutCodeOfConductGuideline(String content) {
+    if (requiredContentPatterns.isEmpty()) {
+      return HAS_REQUIRED_TEXT_IN_CODE_OF_CONDUCT_GUIDELINE.value(true);
+    }
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfo.java:173]: This is what bothered me. If the required patterns list is empty.

Why would we set the `HAS_REQUIRED_TEXT` as `true`?

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfo.java [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+   * @return Content of a code of conduct guideline if found.
+   * @throws IOException If something went wrong.
+   */
+  private Optional<String> lookForCodeOfConductGuideline(GitHubProject project) throws IOException {
+    LocalRepository repository = GitHubDataFetcher.localRepositoryFor(project);
+
+    for (String path : knownCodeofConductGuidelineFiles) {
+      Optional<String> content = repository.readTextFrom(path);
+      if (content.isPresent()) {
+        return content;
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfo.java:157]: looks a bit redundant. As the `readtestFrom()` already returns an Optional if content is available or not.

**Review by sourabhsparkala — COMMENTED** [review]
Everything else looks good. A few comments.

**Comment by sourabhsparkala** [comment]
As discussed please update the docs for the new feature. Thanks

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfo.java (hunk 7 above)
Comment (ajinkyapatil8190) [reply src/main/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfo.java]: Removed this part.

File: docs/getting_oss_rules_of_play_rating.md [hunk 8]
```diff
@@ -18,6 +18,11 @@ The data provider `ContributingGuidelineInfo` allows you to specify 2 types of r
 - The content that you expect must not be part of the contributing guidelines.
 The option is called `requiredContentPatterns` and expects a list of one or multiple Regex. If you enter multiple entries, the contribution guidelines must comply with all the patterns for the rule to pass.
 
+### Code of Conduct Guidelines
+The data provider `CodeOfConductGuidelineInfo` allows you to specify 1 type of regular expression pattern
+- The content that you expect to be part of the code of conduct guidelines.
+The option is called `requiredContentPatterns` and expects a list of one or multiple Regex. If you enter multiple entries, the code of conduct guidelines must comply with all the patterns for the rule to pass.
```
Comment (sourabhsparkala) [inline docs/getting_oss_rules_of_play_rating.md:24]: I think if this is put as parameter un config.yml.

Then the system would not know if it is from Code of Conduct Guidelines or Contributing Guideline Info.

I think we need to change the attribute name in the program and also in the config.yml, test files

File: docs/getting_oss_rules_of_play_rating.md [hunk 9]
```diff
@@ -113,28 +118,30 @@ If everything is correctly set up, the output is going to look like the followin
...
+[+]       Does it have a code of conduct guideline?.................No
+[+]       Does it have a contributing guideline?....................Yes
+[+]       Does it have a license file?..............................Yes
+[+]       Does it have a security policy?...........................Yes
+[+]       Does it have a team with push privileges on GitHub?.......Yes
+[+]       Does it have an admin team on GitHub?.....................Yes
+[+]       Does it have enough admins on GitHub?.....................Yes
+[+]       Does it have enough teams on GitHub?......................Yes
+[+]       Does it have unresolved vulnerability alerts?.............No
+[+]       Does it use an allowed license?...........................No
```
Comment (sourabhsparkala) [inline docs/getting_oss_rules_of_play_rating.md:134]: Shouldn't this be Yes, if this is the same project?

File: docs/getting_oss_rules_of_play_rating.md [hunk 10]
```diff
@@ -113,28 +118,30 @@ If everything is correctly set up, the output is going to look like the followin
...
+[+]       Does the code of conduct guideline have required text?....No
+[+]       Does the contributing guideline have required text?.......Yes
+[+]       Does the license have disallowed content?.................No
+[+]       Is README incomplete?.....................................No
+[+]       Is it compliant with REUSE rules?.........................Yes
+[+]       Is it registered in REUSE?................................Yes
+[+]   Explanation:..Found 1 violated rule
 [+]
-[+] Rating:     9.0  out of 10.0 -> PASSED_WITH_WARNING
+[+] Rating:     0,0  out of 10,0 -> FAILED
```
Comment (sourabhsparkala) [inline docs/getting_oss_rules_of_play_rating.md:144]: This output is not good. The rating value does not look good with 0. Also, it doesn't look good that it returns plain 0. Something maybe wrong here, while calculating the score. Could you check this?

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfo.java [hunk 11]
```diff
@@ -153,9 +153,7 @@ protected ValueSet fetchValuesFor(GitHubProject project) throws IOException {
 
     for (String path : knownCodeofConductGuidelineFiles) {
       Optional<String> content = repository.readTextFrom(path);
-      if (content.isPresent()) {
-        return content;
-      }
+      return content;
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfo.java:156]: I just wanted to know your thoughts if there was a better way. By removing this if condition, it becomes even worse in a for loop.

Now for loop will run only once. Please revert it back.

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfo.java [hunk 12]
```diff
@@ -169,7 +167,7 @@ protected ValueSet fetchValuesFor(GitHubProject project) throws IOException {
    */
   private Value<Boolean> infoAboutCodeOfConductGuideline(String content) {
     if (requiredContentPatterns.isEmpty()) {
-      return HAS_REQUIRED_TEXT_IN_CODE_OF_CONDUCT_GUIDELINE.value(true);
+      return HAS_REQUIRED_TEXT_IN_CODE_OF_CONDUCT_GUIDELINE.value(false);
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfo.java:170]: Again I wanted an explanation as to why?

- If there are no patterns to search and if the file is found. What does this mean?
- If there is a pattern and if you find the file. Does it mean it is the best?
- If there are patterns to search, but not found in the found file. Does it mean its good?
- If there are patterns to search, but no file found. Does it mean the worst?

File: docs/getting_oss_rules_of_play_rating.md (hunk 8 above)
Comment (ajinkyapatil8190) [reply docs/getting_oss_rules_of_play_rating.md:24]: When we create a config object we use the `.class.getSimpleName()`  which return the classname and we append the `.config.yml` to it at the end. We also have 2 different files 

 - CodeOfConductGuidelineInfo.config.yml
 - ContributingGuidelineInfo.config.yml

Which both contain attribute `requiredContentPatterns` so even the attribute name is same the files are different so there shouldn’t be an confusion.

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfo.java (hunk 11 above)
Comment (ajinkyapatil8190) [reply src/main/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfo.java:156]: Reverted.

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfo.java (hunk 12 above)
Comment (ajinkyapatil8190) [reply src/main/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfo.java:170]: - ´ requiredContentPatterns ´ are the patterns we want to check in the file. 
If the patterns are empty that means there are not patterns to be found, which means the file (if present) is perfect and we can move on so - the ´HAS_REQUIRED_TEXT´ is always ´true´ as there is nothing to search for. 
- and if ´ requiredContentPatterns ´ is not empty that means we now need to apply regex and check if the pattern exists in the file's text. which we do later. 
- So the original ´return HAS_REQUIRED_TEXT_IN_CODE_OF_CONDUCT_GUIDELINE.value(true);´ was right.
- I will revert the change.

File: docs/getting_oss_rules_of_play_rating.md (hunk 9 above)
Comment (ajinkyapatil8190) [reply docs/getting_oss_rules_of_play_rating.md:134]: Not sure why the allowed license part is saying no, i tried the same with openui5 repository and for that it also showed No for the ´Does it use an allowed license?´, 
So, I cloned the current master branch and executed the code for openui5 and code-pal-for-abap and for both it showed No for the ´Does it use an allowed license?´
maybe this is an issue on master which is causing this error, as I have only implemented the Code of conduct thing and did not touch any other part of code.

File: docs/getting_oss_rules_of_play_rating.md (hunk 10 above)
Comment (ajinkyapatil8190) [reply docs/getting_oss_rules_of_play_rating.md:144]: Same as above the same thing is happening on the master branch code as well for multiple repositories.

File: src/test/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfoTest.java [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+                    "---\n"
+                  + "requiredContentPatterns:\n"
+                  + "  - \"Contributor Covenant\""));
+
+    ValueSet values = provider.fetchValuesFor(project);
+    checkValue(values, HAS_CODE_OF_CONDUCT, true);
+    checkValue(values, HAS_REQUIRED_TEXT_IN_CODE_OF_CONDUCT_GUIDELINE, true);
+  }
+
+
```
Comment (sourabhsparkala) [inline src/test/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfoTest.java:83]: Please remove the line here

File: src/test/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfoTest.java [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+      CodeOfConductGuidelineInfo provider = new CodeOfConductGuidelineInfo(fetcher);
+      assertEquals(1, provider.requiredContentPatterns().size());
+      assertEquals(
+          "Contributor Covenant",
+          provider.requiredContentPatterns().get(0).pattern());
+    } finally {
+      FileUtils.forceDeleteOnExit(config.toFile());
+    }
+  }
+
```
Comment (sourabhsparkala) [inline src/test/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfoTest.java:103]: Can you add a test here?
- Load CodeOfCOnduct.config.yml
- Load Contributing.config.yml

- Assert when CodeOfConfuct provider has the appropriate Content patterns
- Assert when Contributing provider has the appropriate Content patterns

Also both the patterns will have the same attribute `requiredContentPatterns`

Comment (sourabhsparkala) [reply src/test/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfoTest.java:103]: My mistake. I over-engineered.
I had a look, looks like each Data provider loads a new set of configs required as per the name. Hence this is good for now.

Thank you for the changes.

File: src/test/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfoTest.java (hunk 13 above)
Comment (sourabhsparkala) [reply src/test/java/com/sap/oss/phosphor/fosstars/data/github/CodeOfConductGuidelineInfoTest.java:83]:
```suggestion
```

File: docs/getting_oss_rules_of_play_rating.md (hunk 9 above)
Comment (sourabhsparkala) [reply docs/getting_oss_rules_of_play_rating.md:134]: @ajinkyapatil8190 this is the final part. Please check if you can find a suitable project to generate a report with Rating value > 0

File: docs/getting_oss_rules_of_play_rating.md (hunk 8 above)
Comment (sourabhsparkala) [reply docs/getting_oss_rules_of_play_rating.md]: I checked this. Thanks for the clarification.

File: docs/getting_oss_rules_of_play_rating.md (hunk 10 above)
Comment (sourabhsparkala) [reply docs/getting_oss_rules_of_play_rating.md:144]: https://github.com/SAP/fosstars-rating-core/pull/693#discussion_r714559323

File: docs/getting_oss_rules_of_play_rating.md [hunk 15]
```diff
@@ -73,70 +78,65 @@ In addition, remember to keep the `OssRulesOfPlayRatingMarkdownFormatter.config.
...
 [+] Let's gather info and calculate a rating for:
 [+]   https://github.com/SAP/code-pal-for-abap
-[+] Gathering info about project's README file ...
-[+] Cloning https://github.com/SAP/code-pal-for-abap ...
-[+] Fetching info about project's teams ...
-[+] Gathering info about vulnerability alerts ...
-[+] Gathering info about project's contributing guidelines ...
-[+] Gathering info about project's license ...
-[+] Figuring out how the project uses REUSE ...
-[+] Figuring out if the project has a security policy ...
```
Comment (sourabhsparkala) [inline docs/getting_oss_rules_of_play_rating.md:87]: Please let these be as it is. Also is this the exact flow, which comes in CLI terminal?

**Review by sourabhsparkala — APPROVED** [review]
👍

**Pull request merged**